Added target compactblocks_parse for Bitcoin Core and Rust Bitcoin

yuvicc opened the pull request:
### Added new target `compactblocks_parse` for Bitcoin Core and Rust Bitcoin

see commit message for more details.

Prabhat1308 reviewed (commented):
Seems like a lot of code got touched due to different formatter being used . Would be good to fix that to ease the review process .

i-am-yuvi commented:
> Seems like a lot of code got touched due to different formatter being used . Would be good to fix that to ease the review process .

Agree, will fix that asap!

brunoerg commented:
sorry I pushed to v2, can you rebase please?

brunoerg commented:
Can you please rebase?

i-am-yuvi commented:
> Can you please rebase?

Done!
Can you approve the workflow? @brunoerg

brunoerg commented:
I can review this today but needs rebase.

brunoerg commented:
> > I can review this today but needs rebase.
> 
> Done. Need workflow approval!

Approved.

erickcestari commented on include/bitcoinfuzz/basemodule.cpp:
```diff
@@ -52,4 +52,14 @@ namespace bitcoinfuzz
     {
         return std::nullopt;
     }
+
+    std::optional<std::string> BaseModule::psbt_parse(std::span<const uint8_t> buffer) const
+    {
+        return std::nullopt;
+    }
+
```
Thanks for the PR!

`BaseModule::psbt_parse() ` already exists
```suggestion
```

yuvicc replied:
Done, thanks for this!

brunoerg commented on driver.cpp:
```diff
@@ -202,28 +207,87 @@ namespace bitcoinfuzz
     {
         std::span<const uint8_t> buffer{data, size};
-        if (target == "script") {
+        if (target == "script")
```
I don't think you should include this style changes here.

yuvicc replied:
okay

yuvicc replied:
Done in 8c9a02210ea795d88d33c466d5a59465cac69655

brunoerg commented:
Did you get this code from bitcoinfuzz v1 (master branch)?

brunoerg commented:
Needs rebase.

brunoerg commented:
I think you need to rebase, there is a silent conflict?

i-am-yuvi commented:
> I think you need to rebase, there is a silent conflict?

Done!

brunoerg commented on driver.cpp:
```diff
@@ -226,6 +226,44 @@ namespace bitcoinfuzz
+            {
+                if (*res != *last_response)
+                {
+                    std::cout << "Input cmpt_block (truncated)";
```
No need to print the input, you can remove it.

yuvicc replied:
Done!

brunoerg commented on driver.cpp:
```diff
@@ -32,10 +35,13 @@ namespace bitcoinfuzz
             std::optional<std::vector<bool>> res{module.second->deserialize_block(buffer)};
             if (!res.has_value() || res->empty()) continue;
             if (last_response.has_value()) assert(*res == *last_response);
+            if (!res.has_value() || res->empty()) continue;
```
There are a lot of duplicated code. Should remove.

yuvicc replied:
Thanks for this. Lots of duplicate code came during rebase. I have fixed that!

brunoerg reviewed (changes requested):
In b7761abfc24c47b2ca6db31c46ab44f55754f02d you're adding stuff from core lightning. Please, remove it.

i-am-yuvi commented:
@brunoerg thanks for reviewing, I have made all the changes. Can you approve the workflow!

i-am-yuvi commented:
Not sure why was the CI failing, it was successful on my Ubuntu VM. I have rebased the changes again!

i-am-yuvi commented:
> Not sure why was the CI failing, it was successful on my Ubuntu VM. I have rebased the changes again!

The CI seems to be working fine when I change the order of Compact blocks parse in workflow file to above all the tests steps, see below

- here's the failed job where I haven't changed the order of compact blocks test: https://github.com/i-am-yuvi/bitcoinfuzz/actions/runs/15041239204/job/42273249405

- here I have changed the order, https://github.com/i-am-yuvi/bitcoinfuzz/actions/runs/15041661058/job/42274595943

i-am-yuvi commented:
@brunoerg can you review this! I've rebased this with v2.

brunoerg commented on modules/rustbitcoin/rust_bitcoin_lib/src/lib.rs:
```diff
@@ -193,6 +194,25 @@ pub unsafe extern "C" fn rust_bitcoin_addrv2(data: *const u8, len: usize) -> *mu
+    let res = deserialize_partial::<HeaderAndShortIds>(data_slice);
+
+    match res {
+        Ok(block) => return (block.0.prefilled_txs.len() + block.0.short_ids.len()).try_into().unwrap(),
```
I think we could return more information then simply returning this count. I know this was how we did in bitcoinfuzz v1, but I think we could have a better target here.

yuvicc replied:
Agree, I will look into it!

brunoerg replied:
Cool, move to draft?

yuvicc replied:
Yes, I think this can be moved to draft

yuvicc committed:
added compact blocks parse target for Bitcoin Core

yuvicc committed:
Added target compactblocks_parse for Rust Bitcoin and Embit module for psbt_parsing in workflow file.

yuvicc commented:
Changes since last commit, 3f45d8e4edc116c76583c9890a5e38a47731cbd4: 

- Rebased with v2
- In v1 we were returning the size of total transaction but now we are returning the size of serialized after parsing it, same goes with rustbitcoin as suggested by @brunoerg see [here](https://github.com/bitcoinfuzz/bitcoinfuzz/pull/127#discussion_r2162208462).

Another approach came in my mind was hashing the parsed data, not sure about this approach as I feel returning len would be a simpler and better approach for differential fuzzing.

brunoerg commented on modules/bitcoin/module.cpp:
```diff
@@ -454,5 +457,28 @@ std::optional<std::string> Bitcoin::psbt_parse(std::span<const uint8_t> buffer)
     return result;
 }
 
+std::optional<int> Bitcoin::cmpctblocks_parse(std::span<const uint8_t> buffer) const
```
I don't think it should be `int`, it can be easily be an `unsigned integer`.

brunoerg reviewed (approved):
ACK 3f45d8e4edc116c76583c9890a5e38a47731cbd4

https://github.com/bitcoinfuzz/bitcoinfuzz/pull/127#discussion_r2193125151 is just a nit and can be addressed in a follow-up.
